### PR TITLE
[BUGFIX] uncommitted model should be able to 'becameValid'

### PR DESCRIPTION
**Pull request by yratanov**
Hey guys,

I faced a problem and did a test + added a fix, the simplest way to replicate:
```js
myModel = this.store.createRecord('my-model');
myModel.errors.add('fieldA', ['blank']);
myModel.errors.add('fieldB', ['blank']);
later(() => {
  set(myModel, 'fieldA', 'value');
  set(myModel, 'fieldB', 'value');
}, 1000);
```

It raises an error: 
```
Attempted to handle event `becameValid` on <my-model:null> while in state root.loaded.created.uncommitted.
```

I'm not sure why it is being raised only if there are 2 or more fields with errors.

### CODE REVIEW

**Comment by runspired** [comment]
once the defunct state-machine is dropped this would be a non-issue. Probably the more interesting test to add is what happens to the user-observable state during this transition. You might also want to try completing the work in #7971

**Comment by deanmarano** [comment]
I'm experiencing this issue going from 3.24 to 3.28. @runspired Would it be possible for this to be released as a patch in 3.28?

**Comment by yratanov** [comment]
I think it should be added as a fix for both v3.28 and v4. Even if this code will be dropped completely, at least it will allow users to upgrade. Also, it's a regression.

**Comment by runspired** [comment]
@yratanov I don't think it's a regression, it's simply a state that has never been handled by ember-data. More than likely it should have been handled a long time ago, but never got implemented. If folks are encountering this upgrading from 3.24->3.28 that is especially surprising. It suggests that pre-request-manager activation failed creates were ignored in terms of internal state and would not have added validation errors to individual attributes, which itself would be a bug.

File: packages/-ember-data/tests/integration/model-errors-test.ts [hunk 1]
```diff
@@ -67,4 +70,24 @@ module('integration/model.errors', function (hooks) {
 
     assert.dom('.error-list__error').doesNotExist();
   });
+
+  test('Uncommitted model can become valid after changing 2 erred fields', async function (this: CurrentTestContext, assert) {
+    this.tag = this.owner.lookup('service:store').createRecord('tag');
+    // @ts-ignore
+    const errors = get(this.tag, 'errors');
```
Comment (runspired) [inline packages/-ember-data/tests/integration/model-errors-test.ts:77]: dot notation access should be fine

File: packages/-ember-data/tests/integration/model-errors-test.ts [hunk 2]
```diff
@@ -67,4 +70,24 @@ module('integration/model.errors', function (hooks) {
 
     assert.dom('.error-list__error').doesNotExist();
   });
+
+  test('Uncommitted model can become valid after changing 2 erred fields', async function (this: CurrentTestContext, assert) {
+    this.tag = this.owner.lookup('service:store').createRecord('tag');
```
[the remark this answers is not in the document]
Comment (yratanov) [reply packages/-ember-data/tests/integration/model-errors-test.ts:75]: > we should assert the state name and state flags generated for each part of the transition.

:+1:

> Likely this test should call save ... as manually adding errors is not a well supported or encouraged pattern.

it's a public API and it's the exact case when this issue happens. Also, in the test above we don't save anything.

File: packages/-ember-data/tests/integration/model-errors-test.ts [hunk 3]
```diff
@@ -41,9 +37,16 @@ interface CurrentTestContext {
 module('integration/model.errors', function (hooks) {
   setupRenderingTest(hooks);
 
-  hooks.beforeEach(function (this: CurrentTestContext) {
+  hooks.beforeEach(function () {
     let { owner } = this;
 
+    class ErrorList extends Component<{ model: Model; field: string }> {
```
[the remark this answers is not in the document]
Comment (runspired) [reply packages/-ember-data/tests/integration/model-errors-test.ts:43]: Why?

[the remark this answers is not in the document]
Comment (yratanov) [reply packages/-ember-data/tests/integration/model-errors-test.ts:43]: I assume because it was called in the previous test...

File: packages/-ember-data/tests/integration/model-errors-test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (yratanov) [reply packages/-ember-data/tests/integration/model-errors-test.ts:75]: when you try to save an object with relations (like when using `EmbeddedRecordsMixin`) ember doesn't know how to assign the errors so you have to do it manually.

[the remark this answers is not in the document]
Comment (yratanov) [reply packages/-ember-data/tests/integration/model-errors-test.ts:75]: Even though it's a trick and might not be as great semantically, I think many people did it one way or another (especially since `lid` is not documented anywhere and didn't exist when many apps were created), so why not fix this behaviour. It won't do any harm imo.

[the remark this answers is not in the document]
Comment (runspired) [reply packages/-ember-data/tests/integration/model-errors-test.ts:75]: @yratanov I'm not saying don't make it work :P Just want to ensure we test what we actually support. I guess it makes sense for these edge cases, I respect folks have historically done bad things around transactional saves. Can we get test coverage of both?

File: packages/-ember-data/tests/integration/model-errors-test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (runspired) [reply packages/-ember-data/tests/integration/model-errors-test.ts:43]: It should only need to do this a single time 🤔 I might dig in.

File: packages/-ember-data/tests/integration/model-errors-test.ts [hunk 4]
```diff
@@ -67,4 +70,24 @@ module('integration/model.errors', function (hooks) {
 
     assert.dom('.error-list__error').doesNotExist();
   });
+
+  test('Uncommitted model can become valid after changing 2 erred fields', async function (this: CurrentTestContext, assert) {
+    this.tag = this.owner.lookup('service:store').createRecord('tag');
+    // @ts-ignore
```
[the remark this answers is not in the document]
Comment (runspired) [reply packages/-ember-data/tests/integration/model-errors-test.ts:76]: should figure out removing it, probably by using dot-notation it becomes removeable.

File: packages/-ember-data/tests/integration/model-errors-test.ts (hunk 1 above)
Comment (yratanov) [reply packages/-ember-data/tests/integration/model-errors-test.ts:77]: image

File: packages/-ember-data/tests/integration/model-errors-test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (yratanov) [reply packages/-ember-data/tests/integration/model-errors-test.ts:75]: `currentState` is a private property, can you advise on how to test it here?

On `save`, the record will become `invalid`, therefore it will have `becomeValid` substate. The error appears only for `uncommitted` records.

[the remark this answers is not in the document]
Comment (runspired) [reply packages/-ember-data/tests/integration/model-errors-test.ts:75]: currentState on models isn't private per-say. Though I think we do want to deprecate at least part of what it does (the state name calculation) in favor of folks using the flags only.

I'm going to skip on this for this release cycle as I believe this is probably fixed on master given we've removed the state machine. If it's not lets make tweaks, if it is lets use this as the basis for a fix for older releases and target them.

[the remark this answers is not in the document]
Comment (yratanov) [reply packages/-ember-data/tests/integration/model-errors-test.ts:75]: @runspired it does work in master! Can we use the fix from this branch for 3.28? We can't migrate straight to 4.X right now...

[the remark this answers is not in the document]
Comment (runspired) [reply packages/-ember-data/tests/integration/model-errors-test.ts:75]: Feel free to retarget the PR. Would be great to still get a good test for this onto master though.

**Comment by yratanov** [comment]
closed in favour of #8154

File: packages/-ember-data/tests/integration/model-errors-test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (yratanov) [reply packages/-ember-data/tests/integration/model-errors-test.ts:75]: Opened new pr against `lts-3-28`, I hope it's the correct one? :)
